libraries/AP_Compass/CompassCalibrator.cpp

```diff
@@ -75,139 +76,132 @@ extern const AP_HAL::HAL& hal;
 
 CompassCalibrator::CompassCalibrator()
 {
-    stop();
+    set_status(Status::NOT_STARTED);
 }
 
+// Request to cancel calibration 
 void CompassCalibrator::stop()
 {
-    set_status(Status::NOT_STARTED);
+    request_set_status(Status::NOT_STARTED);
 }
 
-void CompassCalibrator::set_orientation(enum Rotation orientation, bool is_external, bool fix_orientation)
-{
-    _check_orientation = true;
-    _orientation = orientation;
-    _orig_orientation = orientation;
-    _is_external = is_external;
-    _fix_orientation = fix_orientation;
+// Request status cahnge
+void CompassCalibrator::request_set_status(Status status) {
+    WITH_SEMAPHORE(state_sem);
+    _requested_status = status;
+    _status_set_requested = true;
 }
 
-void CompassCalibrator::start(bool retry, float delay, uint16_t offset_max, uint8_t compass_idx)
+void CompassCalibrator::set_tolerance(float tolerance)
 {
-    if (running()) {
+    WITH_SEMAPHORE(state_sem);
+    // Don't do this while we are already started
+    if (cal_state.status != Status::NOT_STARTED) {
         return;
     }
-    _offset_max = offset_max;
-    _attempt = 1;
-    _retry = retry;
-    _delay_start_sec = delay;
-    _start_time_ms = AP_HAL::millis();
-    _compass_idx = compass_idx;
-    set_status(Status::WAITING_TO_START);
+    cal_settings.tolerance = tolerance;
+}
+
+void CompassCalibrator::set_orientation(enum Rotation orientation, bool is_external, bool fix_orientation)
+{
+    WITH_SEMAPHORE(state_sem);
+    cal_settings.check_orientation = true;
+    cal_settings.orientation = orientation;
+    cal_settings.orig_orientation = orientation;
+    cal_settings.is_external = is_external;
+    cal_settings.fix_orientation = fix_orientation;
 }
 
-void CompassCalibrator::get_calibration(Vector3f &offsets, Vector3f &diagonals, Vector3f &offdiagonals, float &scale_factor)
+void CompassCalibrator::start(bool retry, float delay, uint16_t offset_max, uint8_t compass_idx)
 {
-    if (_status != Status::SUCCESS) {
+    if (compass_idx > COMPASS_MAX_INSTANCES) {
         return;
     }
 
-    offsets = _params.offset;
-    diagonals = _params.diag;
-    offdiagonals = _params.offdiag;
-    scale_factor = _params.scale_factor;
+    WITH_SEMAPHORE(state_sem);
+    // Don't do this while we are already started
+    if (cal_state.status == Status::RUNNING_STEP_ONE || 
+       cal_state.status == Status::RUNNING_STEP_TWO) {
+        return;
+    }
+    cal_settings.offset_max = offset_max;
```

Comment:
The changes could be made while not running and thread is still actively reading those values, like while in states `WAITING_FOR_START` etc. I made a change in behaviour in setting by mistake that doesn't allow that, I will fix it.

Reply:
If you look in master right now outside of this PR the calibrator spends no time in `WAITING_FOR_START`. As soon as you set the status to that, the call to set_status immediately calls `set_status(RUNNING_STEP_ONE)` which means by the time it's returned from the first call to `set_status(WAITING_FOR_START)` it's actually in `RUNNING_STEP_ONE`. I admit this is a case where it would actually be easier to factor this state out before the thread split, but the state seems to be unneeded and inducing complexity elsewhere.

Reply:
@WickedShell No, it doesn't always. Check this line https://github.com/ArduPilot/ardupilot/blob/master/libraries/AP_Compass/CompassCalibrator.cpp#L308 it returns false and doesn't set the _statue to `RUNNING_STEP_ONE` before the delay is complete.